src/graphics/program-lib/programs/lit-shader.js

```diff
@@ -0,0 +1,1461 @@
+};
+
+const builtinVaryings = {
+    vVertexColor: "vec4",
+    vPositionW: "vec3",
+    vNormalV: "vec3",
+    vNormalW: "vec3",
+    vTangentW: "vec3",
+    vBinormalW: "vec3",
+    vObjectSpaceUpW: "vec3",
+    vUv0: "vec2",
+    vUv1: "vec2"
+};
+
+const decodeTable = {
+    'rgbm': 'decodeRGBM',
+    'rgbe': 'decodeRGBE',
+    'linear': 'decodeLinear'
+};
+
+class LitShader {
+    constructor(device, options) {
+        this.device = device;
+        this.options = options;
+
+        // resolve custom chunk attributes
+        this.attributes = {
+            vertex_position: SEMANTIC_POSITION
+        };
+
+        if (options.chunks) {
+            this.chunks = {};
+
+            for (const p in shaderChunks) {
+                if (shaderChunks.hasOwnProperty(p)) {
+                    if (options.chunks[p]) {
+                        const chunk = options.chunks[p];
+                        for (const a in builtinAttributes) {
+                            if (builtinAttributes.hasOwnProperty(a) && chunk.indexOf(a) >= 0) {
+                                this.attributes[a] = builtinAttributes[a];
+                            }
+                        }
+                        this.chunks[p] = chunk;
+                    } else {
+                        this.chunks[p] = shaderChunks[p];
+                    }
+                }
+            }
+        } else {
+            this.chunks = shaderChunks;
+        }
+
+        if (options.forceFragmentPrecision &&
+            options.forceFragmentPrecision !== "highp" &&
+            options.forceFragmentPrecision !== "mediump" &&
+            options.forceFragmentPrecision !== "lowp") {
+            options.forceFragmentPrecision = null;
+        }
+
+        if (options.forceFragmentPrecision) {
+            if (options.forceFragmentPrecision === "highp" && device.maxPrecision !== "highp") {
+                options.forceFragmentPrecision = "mediump";
+            }
+            if (options.forceFragmentPrecision === "mediump" && device.maxPrecision === "lowp") {
+                options.forceFragmentPrecision = "lowp";
+            }
+        }
+
+        this.lighting = (options.lights.length > 0) || !!options.dirLightMap || !!options.clusteredLightingEnabled;
+        this.reflections = !!options.reflectionSource;
+        if (!options.useSpecular) options.specularMap = options.glossMap = null;
+        this.shadowPass = options.pass >= SHADER_SHADOW && options.pass <= 17;
+        this.needsNormal = this.lighting || this.reflections || options.ambientSH || options.heightMap || options.enableGGXSpecular ||
+                            (options.clusteredLightingEnabled && !this.shadowPass) || options.clearCoatNormalMap;
+
+        // generated by vshader
+        this.varyings = "";
+        this.vshader = null;
+
+        // supplied by caller
+        this.frontendDecl = null;
+        this.frontendCode = null;
+        this.frontendFunc = null;
+        this.lightingUv = null;
```

Comment:
would be great to have comments on what these are

Reply:
Yeah... I plan to continue working on this in followup PR. This js code interface between frontend and lit-shader will become more formal and documented (along with the options object).